datafusion/physical-plan/src/joins/dynamic_filters.rs

```diff
@@ -0,0 +1,267 @@
+use datafusion_common::{exec_err, DataFusionError};
+use datafusion_expr::Accumulator;
+use datafusion_expr::Operator;
+use datafusion_functions_aggregate::min_max::{MaxAccumulator, MinAccumulator};
+use datafusion_physical_expr::expressions::{BinaryExpr, Column, Literal};
+use datafusion_physical_expr::PhysicalExpr;
+use hashbrown::HashSet;
+use parking_lot::Mutex;
+use std::fmt;
+use std::sync::Arc;
+
+pub struct DynamicFilterInfo {
+    columns: Vec<Arc<Column>>,
+    build_side_names: Vec<String>,
+    inner: Mutex<DynamicFilterInfoInner>,
+}
+
+struct DynamicFilterInfoInner {
+    max_accumulators: Vec<MaxAccumulator>,
+    min_accumulators: Vec<MinAccumulator>,
+    final_expr: Option<Arc<dyn PhysicalExpr>>,
+    batch_count: usize,
+    processed_partitions: HashSet<usize>,
+}
+
+impl fmt::Debug for DynamicFilterInfo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("DynamicFilterInfo")
+            .field("columns", &self.columns)
+            .field("inner", &"<locked>")
+            .finish()
+    }
+}
+
+impl DynamicFilterInfo {
+    // Create a new DynamicFilterInfo instance
+    pub fn try_new(
+        columns: Vec<Arc<Column>>,
+        build_side_names: Vec<String>,
+        data_types: Vec<&DataType>,
+        total_batches: usize,
+    ) -> Result<Self, DataFusionError> {
+        let (max_accumulators, min_accumulators) = data_types
+            .into_iter()
+            .try_fold::<_, _, Result<_, DataFusionError>>(
+                (Vec::new(), Vec::new()),
+                |(mut max_acc, mut min_acc), data_type| {
+                    max_acc.push(MaxAccumulator::try_new(data_type)?);
+                    min_acc.push(MinAccumulator::try_new(data_type)?);
+                    Ok((max_acc, min_acc))
+                },
+            )?;
+
+        Ok(Self {
+            columns,
+            build_side_names,
+            inner: Mutex::new(DynamicFilterInfoInner {
+                max_accumulators,
+                min_accumulators,
+                final_expr: None,
+                batch_count: total_batches,
+                processed_partitions: HashSet::with_capacity(total_batches),
+            }),
+        })
+    }
+
+    // Set the final expression for the filter
+    pub fn with_final_expr(self, expr: Arc<dyn PhysicalExpr>) -> Self {
+        self.inner.lock().final_expr = Some(expr);
+        self
+    }
+
+    // Merge a new batch of data and update the max and min accumulators
+    pub fn merge_batch_and_check_finalized(
+        &self,
+        records: &RecordBatch,
+        partition: usize,
+    ) -> Result<bool, DataFusionError> {
+        let mut inner = self.inner.lock();
+
+        if inner.final_expr.is_some() {
+            return Ok(true);
+        }
+
+        if !inner.processed_partitions.insert(partition) {
+            return Ok(false);
+        }
+
+        inner.batch_count = inner.batch_count.saturating_sub(1);
+        if records.num_rows() == 0 {
```

Comment:
Couldn't the input of a join partition be empty, in which case it will never be finalized?

Reply:
Haven't try that, I'll add a test for that

Reply:
yes, it happens sometime, haven't check the reason

Reply:
I added a new judge "if a partition id has been recorded and it return empty batch again, we treat it as an empty partition and return true"